File: core/src/com/bosowski/oslark/generation/areas/DungeonCell.java
package com.bosowski.oslark.generation.areas;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.bosowski.oslark.World;
import com.bosowski.oslark.enums.Direction;
import com.bosowski.oslark.gameObjects.Terrain;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.ThreadLocalRandom;

public class DungeonCell extends Terrain {

    HashMap<Direction, Terrain> walls = new HashMap<>();
    private static final float chanceOfDifferentTexture = 0.03f;

    public DungeonCell(String name, Vector3 position, boolean collides) {
        super(ThreadLocalRandom.current().nextFloat() <= chanceOfDifferentTexture ? "floor"+ThreadLocalRandom.current().nextInt(4,11) : "floor4", position, collides);
    }

    public void instantiate() {
        World.instance.instantiate(this);
        for (Terrain wall : walls.values()) {
            World.instance.instantiate(wall);
        }
    }

    public Vector2 getVector2() {
        return new Vector2(position.x, position.y);
    }

    public void setUpWalls(HashMap<Vector2, DungeonCell> otherCells) {
        for (Direction direction : Direction.getDirections()) {
            if (!otherCells.containsKey(getVector2().add(direction.value))) {
                addWall(direction);
            }
        }
    }

    public void clear() {
        for (Terrain wall : walls.values()) {
            World.instance.destroy(wall);
        }
        walls.clear();
        World.instance.destroy(this);
    }

    public ArrayList<DungeonCell> getNeighbours(HashMap<Vector2, DungeonCell> otherCells) {
        ArrayList<DungeonCell> neighbours = new ArrayList<>();
        for (Direction direction : Direction.getDirections()) {
            Vector2 neighbourPosition = getVector2().add(direction.value);
            if (otherCells.containsKey(neighbourPosition)) {
                neighbours.add(otherCells.get(neighbourPosition));
            }
        }
        return neighbours;
    }

    private void addWall(Direction direction) {
        Terrain wall;

        if (direction == Direction.LEFT) {
            wall = new Terrain("wallLeft", new Vector3(position.x, position.y, 0), true);
            wall.getCollisionBox().y = wall.getCollisionBox().y - wall.getOrigin().y;
            wall.setDimension(new Vector2(0.2f, 2f));
            wall.getCollisionBox().width = wall.getDimension().x;
            wall.setOrigin(new Vector2(0.5f, 0.5f));
            walls.put(Direction.LEFT, wall);
        } else if (direction == Direction.RIGHT) {
            wall = new Terrain("wallRight", new Vector3(position.x, position.y, 0), true);
            wall.setDimension(new Vector2(-0.2f, 2f));
            wall.getCollisionBox().x = wall.getPosition().x + wall.getOrigin().x;
            wall.getCollisionBox().y = wall.getPosition().y - wall.getOrigin().y;
            wall.getCollisionBox().width = wall.getDimension().x;
            wall.setOrigin(new Vector2(-0.5f, 0.5f));
            walls.put(Direction.RIGHT, wall);
        } else if (direction == Direction.DOWN) {
            wall = new Terrain("wallDown", new Vector3(position.x, position.y, 0), true);
            wall.getCollisionBox().y = wall.getCollisionBox().y - wall.getOrigin().y;
            wall.getCollisionBox().height = 0.1f;
            walls.put(Direction.DOWN, wall);
        } else {
            float chance = ThreadLocalRandom.current().nextFloat();
            int wallType = 4;
            if(chance <= chanceOfDifferentTexture){
                wallType = ThreadLocalRandom.current().nextInt(0, 11);
            }
            if (wallType < 4) {
                name = "floor" + wallType;
                setUp();
            }
            wall = new Terrain("wallUp" + wallType, new Vector3(position.x, position.y + 1f, 0), true);
            if (wallType < 2) {
                wall.setDimension(new Vector2(1f, 1.117f));
            } else if (wallType == 2) {
                wall.setDimension(new Vector2(1f, 1.495f));
            }
            wall.getCollisionBox().y = wall.getCollisionBox().y - wall.getOrigin().y;
            walls.put(Direction.UP, wall);
        }
    }
}

File: OslarkGame/core/src/com/bosowski/oslark/generation/DungeonCell.kt
package com.bosowski.oslark.generation

import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.BodyDef
import com.badlogic.gdx.physics.box2d.EdgeShape
import com.bosowski.oslark.components.TextureComponent
import com.bosowski.oslark.enums.Direction
import com.bosowski.oslark.gameObjects.GameObject
import com.bosowski.oslark.Assets
import com.bosowski.oslark.components.AnimationComponent
import com.bosowski.oslark.components.ColliderComponent
import com.bosowski.oslark.utils.Util
import java.util.*

class DungeonCell(position: Vector2, private val random: Random, val index: Int): Generation{

  val cell = GameObject(position,-1, bodyType = BodyDef.BodyType.StaticBody)
  private val chanceOfDifferentFloor = 0.25f

  private val walls = HashMap<Direction, GameObject>()

  init{
    val floorType: Int = if(random.nextFloat() <= chanceOfDifferentFloor){
      Util.randomInt(random,4,11)
    }
    else{
      4
    }
    val textureComponent = TextureComponent(Assets.textures["floor$floorType"]!!)
    cell.addComponent(textureComponent)
  }

  fun instantiate() {
    cell.instantiate()
    for (wall in walls.values) {
      wall.instantiate()
    }
  }

  fun setUpWalls(otherCells: HashMap<Vector2, DungeonCell>) {
    for (direction in Direction.directions) {
      if (!otherCells.containsKey(Vector2(cell.transform.position).add(direction.value))) {
        addWall(direction)
      }
    }
  }

  override fun clear() {
    for (wall in walls.values) {
      wall.destroy()
    }
    walls.clear()
    cell.destroy()
  }

  fun getNeighbours(otherCells: HashMap<Vector2, DungeonCell>): ArrayList<DungeonCell> {
    val neighbours = ArrayList<DungeonCell>()
    for (direction in Direction.directions) {
      val neighbourPosition = Vector2(cell.transform.position).add(direction.value)
      if (otherCells.containsKey(neighbourPosition)) {
        neighbours.add(otherCells[neighbourPosition]!!)
      }
    }
    return neighbours
  }

  private fun addWall(direction: Direction) {
    walls[direction] = getWall(cell, direction, random)
  }

  private fun getWall(cell: GameObject, direction: Direction, random: Random): GameObject {

    val chanceOfDifferentWall = 0.05f

    var wall = GameObject(cell.transform.position, bodyType = BodyDef.BodyType.StaticBody)
    when (direction) {
      Direction.LEFT -> {
        val textureComponent = TextureComponent(Assets.textures["wallLeft"]!!)
        textureComponent.origin = Vector2(0.5f, 0.0f)
        textureComponent.dimension = Vector2(0.2f, 2f)
        wall.addComponent(textureComponent)
        //type = BodyDef.BodyType.StaticBody, centre = Vector2(- textureComponent.origin.x + 0.1f, 0f), width = 0.2f, height = 1f
        val shape = EdgeShape()
        shape.set(Vector2(-textureComponent.origin.x + textureComponent.dimension.x, -0.0f), Vector2(-textureComponent.origin.x + textureComponent.dimension.x, 1f))
        val colliderComponent = ColliderComponent(shape, 0f)
        wall.addComponent(colliderComponent)
        return wall
      }
      Direction.RIGHT -> {
        val textureComponent = TextureComponent(Assets.textures["wallRight"]!!)
        textureComponent.origin = Vector2(-0.5f, 0.0f)
        textureComponent.dimension = Vector2(0.2f, 2f)
        wall.addComponent(textureComponent)
        //type = BodyDef.BodyType.StaticBody, centre = Vector2( - textureComponent.origin.x, 0f), width = 0.2f, height = 1f
        val shape = EdgeShape()
        shape.set(Vector2(-textureComponent.origin.x, -0.0f), Vector2(-textureComponent.origin.x, 1f))
        val colliderComponent = ColliderComponent(shape, 0f)
        wall.addComponent(colliderComponent)
        return wall
      }
      Direction.DOWN -> {
        val textureComponent = TextureComponent(Assets.textures["wallDown"]!!)
        textureComponent.origin = Vector2(0.5f, 0.0f)
        wall.addComponent(textureComponent)
        //type = BodyDef.BodyType.StaticBody, centre = Vector2(0f,  - textureComponent.origin.y), width = 1f, height = 0.01f
        val shape = EdgeShape()
        shape.set(Vector2(-textureComponent.origin.x, -textureComponent.origin.y), Vector2(textureComponent.origin.x, -textureComponent.origin.y))
        val colliderComponent = ColliderComponent(shape, 0f)
        wall.addComponent(colliderComponent)
        return wall
      }
      else -> {
        wall = GameObject(Vector2(cell.transform.position.x, cell.transform.position.y + 1), bodyType = BodyDef.BodyType.StaticBody)
        val chance = random.nextFloat()
        var wallType = 4
        if (chance <= chanceOfDifferentWall) {
          wallType = Util.randomInt(random, 0, 10)
        }
        val wallImage: TextureComponent
        if (wallType < 2) {
          cell.removeComponent("TextureComponent")
          cell.addComponent(AnimationComponent(Assets.animations["floor$wallType"]!!))
          wallImage = AnimationComponent(Assets.animations["wallUp$wallType"]!!)
          wall.addComponent(wallImage)
        } else {
          wallImage = TextureComponent(Assets.textures["wallUp$wallType"]!!)
          wall.addComponent(wallImage)
        }

        if (wallType in 2..3) {
          cell.addComponent(TextureComponent(Assets.textures["floor$wallType"]!!))
        }
        if (wallType < 2) {
          wallImage.dimension = Vector2(1f, 1.117f)
        } else if (wallType == 2) {
          wallImage.dimension = Vector2(1f, 1.495f)
        }
        //type = BodyDef.BodyType.StaticBody, centre = Vector2(0f, -wallImage.origin.y), width = 1f, height = 0.01f
        val shape = EdgeShape()
        shape.set(Vector2(-wallImage.origin.x, -wallImage.origin.y), Vector2(wallImage.origin.x, -wallImage.origin.y))
        val colliderComponent = ColliderComponent(shape, 0f)
        wall.addComponent(colliderComponent)
        return wall
      }
    }
  }
}
